iree/tools/utils/cpu_features.c

```diff
@@ -0,0 +1,165 @@
+      return iree_cpu_features_aarch64_read_ID_AA64ISAR0_EL1();
+    case iree_cpu_features_aarch64_register_ID_AA64ISAR1_EL1:
+      return iree_cpu_features_aarch64_read_ID_AA64ISAR1_EL1();
+    case iree_cpu_features_aarch64_register_count:  // case, not default, so
+      break;  // the compiler can warn if we forget an enum value.
+  }
+  assert(false && "bad iree_cpu_features_aarch64_register_t value");
+  return 0;
+}
+
+// Returns the value of a cpu feature register. Caches the value in the
+// iree_cpu_features_t struct.
+static uint64_t iree_cpu_features_aarch64_get_register(
+    iree_cpu_features_t* features, iree_cpu_features_aarch64_register_t idx) {
+  if (!features->registers_initialized[idx]) {
+    if (iree_devices_features_aarch64_register_access()) {
+      features->registers[idx] = iree_cpu_features_aarch64_read_register(idx);
+    } /* else: leave registers[idx] with its initial 0 bits */
+    features->registers_initialized[idx] = true;
+  }
+  return features->registers[idx];
+}
+
+// Returns true if +dotprod is supported.
+static bool iree_cpu_features_aarch64_dotprod(iree_cpu_features_t* features) {
+  uint64_t register_value = iree_cpu_features_aarch64_get_register(
+      features, iree_cpu_features_aarch64_register_ID_AA64ISAR0_EL1);
+  return (register_value >> 44) & 1;
+}
+
+// Returns true if +i8mm is supported.
+static bool iree_cpu_features_aarch64_i8mm(iree_cpu_features_t* features) {
+  uint64_t register_value = iree_cpu_features_aarch64_get_register(
+      features, iree_cpu_features_aarch64_register_ID_AA64ISAR1_EL1);
+  return (register_value >> 52) & 1;
+}
+
+typedef iree_cpu_features_t iree_cpu_features_t;
+
+#else  // not defined IREE_ARCH_ARM_64
+
+// Not-implemented case. Decided in PR #8316 to make it non-empty just to avoid
+// edge cases with empty structs.
+struct iree_cpu_features_t {
+  int unused;
+};
+
+#endif  // defined IREE_ARCH_ARM_64
+
+iree_status_t iree_cpu_features_create(iree_allocator_t allocator,
+                                       iree_cpu_features_t** cpu_features) {
+  return iree_allocator_malloc(allocator, sizeof(iree_cpu_features_t),
+                               (void**)cpu_features);
+}
+
+void iree_cpu_features_destroy(iree_allocator_t allocator,
+                               iree_cpu_features_t* cpu_features) {
+  iree_allocator_free(allocator, cpu_features);
+}
+
+iree_status_t iree_cpu_features_query(iree_cpu_features_t* cpu_features,
+                                      iree_string_view_t feature,
+                                      bool* result) {
+  *result = false;
+
+#ifdef IREE_ARCH_ARM_64
+  if (iree_string_view_equal(feature, iree_make_cstring_view("+dotprod"))) {
+    *result = iree_cpu_features_aarch64_dotprod(cpu_features);
+    return iree_ok_status();
+  }
+  if (iree_string_view_equal(feature, iree_make_cstring_view("+i8mm"))) {
+    *result = iree_cpu_features_aarch64_i8mm(cpu_features);
+    return iree_ok_status();
+  }
+#endif
```

Comment:
when possible add the condition to the #endif to make nesting clearer - easiest to always just use the same condition for simple things but possible to widen it if it makes it more readable (`#if FOO_A || FOO_B && !FOO_C ... #endif  // FOO_* compatibility check` etc)
```suggestion
#endif  // IREE_ARCH_ARM_64
```

Reply:
ok